### ISSUE TITLE
[4.x]: craft asset query order by size not working

### ISSUE THREAD

**Issue by jankohlbach** [craftcms/cms#12652]
### What happened?

for some reason I cannot query assets ordered by size, it's just as it wouldn't consider it actually, not ordered at all, both in frontend query and also in the backend for the volumes
every other order works, by name, by date, works, but not by size
anyone an idea what this could be, don't find any hint online, please help
also I looked in the database, there the size looks correct and there I also can sort the table by size, so I don't know why craft can't

### Craft CMS version

4.3.x

### PHP version

8.1

### Installed plugins and versions

-

**Comment by brianjhanson** [comment]
Thanks for reaching out! Unfortunately, I'm having trouble replicating this. When sorting from twig with the following below things seem to work correctly. Do you have any custom code that might be causing issues?


```twig
{% set assets = craft.assets().orderBy('width desc').all() %}

{% for asset in assets %}
    <div>
        <h3>{{ asset.title }}</h3>
        {{ asset.width }} x {{ asset.height }}
    </div>
{% endfor %}
```

As for the CP, `width` & `height` aren't default sorting options. I've opened a PR to address this, but in the mean time you're able to add them yourselves via a custom module and the following code. 


```php
<?php

use craft\elements\Asset;
use craft\events\RegisterElementSortOptionsEvent;
use craft\helpers\ArrayHelper;
use yii\base\Event;

Event::on(
    Asset::class,
    Asset::EVENT_REGISTER_SORT_OPTIONS,
    function(RegisterElementSortOptionsEvent $event) {
        $event->sortOptions = ArrayHelper::merge($event->sortOptions, [
            [
                'label' => Craft::t('app', 'Width'),
                'orderBy' => 'width',
                'attribute' => 'width',
            ],
            [
                'label' => Craft::t('app', 'Height'),
                'orderBy' => 'height',
                'attribute' => 'height',
            ],
        ]);
    }
);
```

**Comment by brianjhanson** [comment]
Thanks for the detail! Sorting by `size` also seems to be working for me. Is there anything in the `storage/logs` that might provide a clue?

**Comment by jankohlbach** [comment]
not really, the only thing I can find is the log of the normal request, it also has the correct parameter set order by size
is there something different in the orderBy function? I cannot find it in the craft code, but is it maybe something like stopping if the uploaderId is null or something
maybe there's a condition or special block of code that causes the sorting to stop because of some not normal state in our db

**Comment by jankohlbach** [comment]
I'm pretty sure it's because of our state of the db and assets, volumes etc. but I don't know how to debug it or find out what exactly the issue is

**Comment by brianjhanson** [comment]
Can you send a database dump along with your `composer.lock` and `composer.json` files over to `support@craftcms.com` and we can take a look?

**Comment by jankohlbach** [comment]
sent you an email, thanks in advance

**Comment by brianjhanson** [comment]
Thanks! One thing I just thought of, do you have a custom field with a handle of `size` on your assets at all?

**Comment by jankohlbach** [comment]
not that I would know of
this project has been a long (tough) process, so if you see anything we messed up or could improve in this way don't hesitate ...

**Comment by brianjhanson** [comment]
Thanks for the extra artifacts. It does look like you have another field called `size` that's getting used for the order instead of the asset `size` column. I'm still working through how best to fix or work around this, but I wanted to give you an update in the mean time.

**Comment by jankohlbach** [comment]
ok, that's very interesting and indeed explains the issue, thanks for that already
maybe it comes from image transforms? or from any other plugin?
we're hosting on servd btw

**Comment by jankohlbach** [comment]
btw. we changed the handle of that size field and it solved the issue
maybe it would be a good idea to prevent field-handles to be called like that? I think I remember us having that issue before because we name a field-handle with handle or something and another functionality broke
maybe it's an idea to have reserved words like 'handle' or 'size' that cannot be used as field-handles

**Comment by brianjhanson** [comment]
We already have a few lists of reserved words to prevent this kind of thing, but in your case, it wouldn't have helped. 

The `orderBy` param will build a key => value map of all fields in the content table when being applied. The order of operations there means that the `size` orderBy param ends up getting set to the column name from the `content` table rather than the `assets` table, which causes the sort to be incorrect. 

On the front end, you can get around this by doing, which will order things properly. 

```twig
{% set assets = craft.assets()
    .orderBy('assets.size desc')
    .all()
%}
```

I'm still thinking about the best place to fix this for CP sorts though.

**Comment by brandonkelly** [comment]
This has been fixed for Craft 4.4 via 74211f8637db0e5fa661d351df98c915c7b94058. In the meantime, avoid using `size` as a custom field handle.

**Comment by brandonkelly** [comment]
Craft 3.8 and 4.4 are out with `width` and `height` sorting (and fix for `size` in 4.4).
